### PR TITLE
Creation of the build_demand_profiles script where demand_profiles.csv is generated

### PR DESCRIPTION
**Pull request by DeniseGiub**
## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

File: Snakefile [hunk 1]
```diff
@@ -341,6 +341,29 @@ if config["enable"].get("retrieve_cost_data", True):
             move(input[0], output[0])
 
 
+rule build_demand:
```
Comment (pz-max) [inline Snakefile]: Maybe `build_demand_profile` to make in coherent with `build_renewable_profiles`

File: Snakefile [hunk 2]
```diff
@@ -341,6 +341,29 @@ if config["enable"].get("retrieve_cost_data", True):
...
+    input:
+        elec_network="networks/" + RDIR + "elec.nc",
```
Comment (pz-max) [inline Snakefile]: elec_network can be removed

File: Snakefile [hunk 3]
```diff
@@ -341,6 +341,29 @@ if config["enable"].get("retrieve_cost_data", True):
...
+        regions="resources/" + RDIR + "bus_regions/regions_onshore.geojson",
+        load=load_data_paths,
+        #gadm_shapes="resources/" + RDIR + "shapes/MAR2.geojson", 
+        #using this line instead of the following will test updated gadm shapes for MA.
+        #To use: downlaod file from the google drive and place it in resources/" + RDIR + "shapes/
+        #Link: https://drive.google.com/drive/u/1/folders/1dkW1wKBWvSY4i-XEuQFFBj242p0VdUlM
+        gadm_shapes="resources/" + RDIR + "shapes/gadm_shapes.geojson",
+    output:
+        "networks/" + RDIR + "elec_1.nc",
```
Comment (pz-max) [inline Snakefile]: Output would be ideally a `csv file`

So in a way, what goes in is the GEGIS hourly time-series for a region e.g. Africa (which has hourly data for each country). What is the expected output is then a `csv` with the `load` distributed across the regions within a country. Debug the variable `load` from the below function which should be contained in the csv.

`n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)`

This function should remain in `add_electricity`

**Review by pz-max — CHANGES_REQUESTED** [review]

File: scripts/build_demand.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            # (refer to vresutils.load._upsampling_weights)
+            # TODO: require adjustment for Africa
+            factors = normed(0.6 * normed(gdp_n) + 0.4 * normed(pop_n))
+            return pd.DataFrame(
+                factors.values * l.values[:, np.newaxis],
+                index=l.index,
+                columns=factors.index,
+            )
+
+    load = pd.concat(
```
Comment (pz-max) [inline scripts/build_demand.py]: Export new load as `csv`.. recommend storing it in `resources/demand.csv`

File: scripts/build_demand.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        [
+            upsample(cntry, group)
+            for cntry, group in regions.geometry.groupby(regions.country)
+        ],
+        axis=1,
+    )
+
+    n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)
```
Comment (pz-max) [inline scripts/build_demand.py]: I prefer this option: Maybe the `substation_lv_i` information can already be stored in `load` (which required then elec.nc in build_demand)

Then you could write in `add_electricity` and add the load like
`n.madd("Load", load.index, bus=load.index, p_set=load)`  (or load.columns depending what has the information.. probably the index has the `snapshot` information and columns the `substation_lv_i` information)

File: Snakefile (hunk 1 above)
Comment (DeniseGiub) [reply Snakefile]: Done!

File: scripts/build_demand.py (hunk 4 above)
Comment (DeniseGiub) [reply scripts/build_demand.py]: Done

File: Snakefile (hunk 3 above)
Comment (DeniseGiub) [reply Snakefile]: Done, the demand csv has the snapshots as rows and buses as columns.

File: scripts/build_demand.py (hunk 5 above)
Comment (DeniseGiub) [reply scripts/build_demand.py]: This was done partially, meaning that the attach_load function now is back in add_electricity and kind of works, it just needs some adjustments, which I will do tomorrow :)

File: scripts/build_demand_profile.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    admin_shapes = snakemake.input.gadm_shapes
+    scale = snakemake.config["load_options"]["scale"]
+    start_date = snakemake.config["snapshots"]["start"]
+    end_date = snakemake.config["snapshots"]["end"]
+
+    create_demand_profile(
+        n, load_paths, regions, admin_shapes, countries, scale, start_date, end_date
+    )
+
+    n.meta = snakemake.config
```
Comment (pz-max) [inline scripts/build_demand_profile.py]: Probably no need to store the config here.
The function will output a csv not a network

File: scripts/build_demand_profile.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        os.chdir(os.path.dirname(os.path.abspath(__file__)))
+        snakemake = mock_snakemake("build_demand_profile")
+        sets_path_to_root("pypsa-earth")
+    configure_logging(snakemake)
+
+    n = pypsa.Network(snakemake.input.base_network)
+
+    # Snakemake imports:
+    regions = snakemake.input.regions
+
```
Comment (pz-max) [inline scripts/build_demand_profile.py]: empty row can be removed

File: scripts/build_demand_profile.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        axis=1,
+    )
+
+    start_date = pd.to_datetime(start_date)
+    end_date = pd.to_datetime(end_date)- pd.Timedelta(hours=1)
+    selected_load = load.loc[start_date:end_date]
+    print(selected_load)
+    selected_load.to_csv("resources/demand.csv", header=True)
+
+    # n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)
```
Comment (pz-max) [inline scripts/build_demand_profile.py]: remove unnecessary comments

File: scripts/build_demand_profile.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            upsample(cntry, group)
+            for cntry, group in regions.geometry.groupby(regions.country)
+        ],
+        axis=1,
+    )
+
+    start_date = pd.to_datetime(start_date)
+    end_date = pd.to_datetime(end_date)- pd.Timedelta(hours=1)
+    selected_load = load.loc[start_date:end_date]
+    print(selected_load)
```
Comment (pz-max) [inline scripts/build_demand_profile.py]: Probably no print of whole dataframe necessary.
Use logger for some stats e.g. `logger.info(f"Load data scaled with scalling factor {scale}.")`

File: scripts/build_demand_profile.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
+# -*- coding: utf-8 -*-
+# SPDX-FileCopyrightText: : 2017-2020 The PyPSA-Eur Authors, 2021 PyPSA-Africa Authors
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+# coding: utf-8
+"""
+Adds load to a network having generators and existing hydro storage units.
```
Comment (pz-max) [inline scripts/build_demand_profile.py]: The whole docstring needs changes

File: scripts/build_demand.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+# SPDX-FileCopyrightText: : 2017-2020 The PyPSA-Eur Authors, 2021 PyPSA-Africa Authors
```
Comment (pz-max) [inline scripts/build_demand.py]: need to delete `build_demand.py`

File: scripts/add_electricity.py [hunk 12]
```diff
@@ -883,8 +961,16 @@ def add_nice_carrier_names(n, config):
...
     )
+
     attach_hydro(n, costs, ppl)
 
+    demand_profile = snakemake.input["demand_profile"]
+    regions = snakemake.input.regions
+
+    attach_load(n, demand_profile, regions)
+
+    load_paths = snakemake.input["load"]
```
Comment (pz-max) [inline scripts/add_electricity.py]: load_paths necessary? seems it can be removed

File: scripts/add_electricity.py [hunk 13]
```diff
@@ -861,7 +940,6 @@ def add_nice_carrier_names(n, config):
 
     conventional_carriers = snakemake.config["electricity"]["conventional_carriers"]
 
-    attach_load(n, load_paths, regions, admin_shapes, countries, scale)
```
Comment (pz-max) [inline scripts/add_electricity.py:859]: Keep reading here `attach_load()` (because PyPSA-Eur does it here as well -- good for comparisons of scripts)

File: scripts/add_electricity.py [hunk 14]
```diff
@@ -883,8 +961,16 @@ def add_nice_carrier_names(n, config):
         extendable_carriers,
         snakemake.config["lines"]["length_factor"],
     )
+
     attach_hydro(n, costs, ppl)
 
+    demand_profile = snakemake.input["demand_profile"]
+    regions = snakemake.input.regions
```
Comment (pz-max) [inline scripts/add_electricity.py]: Keep reading the inputs at the beginning of the __main__ (like at the location of the previous 830-832 that you removed)
We currently always read all the inputs shortly after the __main__ to have a good overview

File: scripts/add_electricity.py [hunk 15]
```diff
@@ -290,65 +373,64 @@ def attach_load(
...
-    load = pd.concat(
-        [
-            upsample(cntry, group)
-            for cntry, group in regions.geometry.groupby(regions.country)
-        ],
-        axis=1,
-    )
-
-    n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)
+    # substation_lv_i = n.buses.index[n.buses["substation_lv"]]
```
Comment (pz-max) [inline scripts/add_electricity.py]: Remove instead of commenting out

File: scripts/add_electricity.py [hunk 16]
```diff
@@ -266,17 +266,100 @@ def get_load_paths_gegis(ssp_parentfolder, config):
...
+
+    demand_df = pd.read_csv(demand_profile, index_col=0)
+    #     #demand_df = demand_df.set_index([n.snapshots]).rename_axis("time")
+    #     a=14
+    #     #for i in len(demand_df.columns):
+    #     n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)
+    for i in demand_df.columns:
+
+        n.madd("Load", demand_df.columns, bus=demand_df.columns, p_set=[i])
+    a=15
```
Comment (pz-max) [inline scripts/add_electricity.py]: do we need here a for-loop? previously we read the whole dataframe like:
`n.madd("Load", substation_lv_i, bus=substation_lv_i, p_set=load)`

Comment (pz-max) [reply scripts/add_electricity.py]: a=15 seems also to be misplaced

File: Snakefile [hunk 17]
```diff
@@ -341,6 +341,29 @@ if config["enable"].get("retrieve_cost_data", True):
             move(input[0], output[0])
 
 
+rule build_demand_profile:
```
Comment (pz-max) [inline Snakefile]: sorry, my bad. When I mean coherent `build_demand_profile` is not correct. It should be `rule build_demand_profiles:`

Also, we usually name the function as the rule. So it would be `"scripts/build_demand_profiles.py"`

File: scripts/build_demand.py (hunk 4 above)
Comment (pz-max) [reply scripts/build_demand.py]: probably call it `demand_profiles` to make it consistent :+1:

File: scripts/build_demand.py (hunk 5 above)
Comment (pz-max) [reply scripts/build_demand.py]: Can we not get rid of the `substation_lv_i` stuff in `add_electricity.py`?
It would be great if all that stuff is dealt with in `build_demand_profiles.py` and that the `demand_profiles.csv` contains all the information such that `substation_lv_i` is not needed.

**Review by pz-max — CHANGES_REQUESTED** [review]
Hi @DeniseGiub , just had some time at the airport to check the changes that you made.
We are on a good way & I know that you would have fixed some of the comments that I added as reminder. Looking forward to review the PR soon :)

File: Snakefile [hunk 18]
```diff
@@ -416,15 +439,18 @@ rule add_electricity:
...
+        #load=load_data_paths,
         #gadm_shapes="resources/" + RDIR + "shapes/MAR2.geojson", 
         #using this line instead of the following will test updated gadm shapes for MA.
         #To use: downlaod file from the google drive and place it in resources/" + RDIR + "shapes/
         #Link: https://drive.google.com/drive/u/1/folders/1dkW1wKBWvSY4i-XEuQFFBj242p0VdUlM
         gadm_shapes="resources/" + RDIR + "shapes/gadm_shapes.geojson",
         hydro_capacities="data/hydro_capacities.csv",
+        demand_profile="resources/demand.csv",
+        regions="resources/" + RDIR + "bus_regions/regions_onshore.geojson",
+        load=load_data_paths,
```
Comment (davide-f) [inline Snakefile]: Probably in add_electricity demand.csv is the only input needed: if the substation_lv is true for a bus name X, then demand csv will have a column named coherently to the bus X. If it is false, no column will be added and that is enough.
What do you think?
Needs to be checked if regions is only needed for the demand and nothing else.

load=load_data_paths may be removed

Moreover, in general do not move rows of inputs unless necessary (e.g. load and regions)

P.S maybe the name "demand.csv" may be revised into "demand_profiles.csv" coherently with renewable profiles).

Comment (davide-f) [reply Snakefile]: Max has already added several comments that address most of the parts

File: scripts/build_demand_profile.py (hunk 6 above)
Comment (DeniseGiub) [reply scripts/build_demand_profile.py]: Deleted!

File: scripts/build_demand_profile.py (hunk 8 above)
Comment (DeniseGiub) [reply scripts/build_demand_profile.py]: Yes

File: scripts/build_demand_profile.py (hunk 9 above)
Comment (DeniseGiub) [reply scripts/build_demand_profile.py]: Sorry I was doing tries, now it's removed.

File: Snakefile (hunk 18 above)
Comment (DeniseGiub) [reply Snakefile]: I eventually got rid of substations_lv_i, not needed anymore. I also applied all the other comments. Thanks to you and to Max for reviewing :)

File: scripts/add_electricity.py [hunk 19]
```diff
@@ -297,58 +291,9 @@ def attach_load(
         Now attached with load time series
     """
     substation_lv_i = n.buses.index[n.buses["substation_lv"]]
```
Comment (davide-f) [inline scripts/add_electricity.py]: This line is no more needed right?

File: scripts/build_demand.py (hunk 11 above)
Comment (davide-f) [reply scripts/build_demand.py]: and also build_demand profile.py

File: scripts/build_demand_profiles.py [hunk 20]
```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+# SPDX-FileCopyrightText: : 2017-2020 The PyPSA-Eur Authors, 2021 PyPSA-Africa Authors
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+# coding: utf-8
+"""
+Creates an electric demand file.
```
Comment (davide-f) [inline scripts/build_demand_profiles.py]: Comment on docstring by max applies here.
Maybe good to add more details for example on what input (e.g. gegis) and output data.

Comment (davide-f) [reply scripts/build_demand_profiles.py]: Moreover, this file needs to be added to the API documentation.
Need to modify file doc/api_reference.rst and add some lines alike the other scripts/rules

File: scripts/simplify_network.py [hunk 21]
```diff
@@ -48,7 +48,7 @@
 - ``resources/costs.csv``: The database of cost assumptions for all included technologies for specific years from various sources; e.g. discount rate, lifetime, investment (CAPEX), fixed operation and maintenance (FOM), variable operation and maintenance (VOM), fuel costs, efficiency, carbon-dioxide intensity.
 - ``resources/regions_onshore.geojson``: confer :ref:`busregions`
 - ``resources/regions_offshore.geojson``: confer :ref:`busregions`
-- ``networks/elec.nc``: confer :ref:`electricity`
+- ``networks/elec_1.nc``: confer :ref:`electricity`
```
Comment (davide-f) [inline scripts/simplify_network.py]: Revert this change

File: scripts/add_electricity.py (hunk 19 above)
Comment (DeniseGiub) [reply scripts/add_electricity.py]: yes

File: scripts/simplify_network.py (hunk 21 above)
Comment (DeniseGiub) [reply scripts/simplify_network.py]: Done :)

File: scripts/build_demand_profiles.py (hunk 20 above)
Comment (DeniseGiub) [reply scripts/build_demand_profiles.py]: I modified the api_reference. I also improved the docstring in build_demand_profiles, hope it's good now.

**Review by davide-f — APPROVED** [review]
Looks good to me :)

As a little comment; have you verified locally that the results of the new version are equivalent to the ones we had before?
Just comparing the output objective value is enough.
The reports in the CI could be used as comparison

Nice contribution! Let's wait for @pz-max

File: scripts/add_electricity.py [hunk 22]
```diff
@@ -266,14 +257,8 @@ def get_load_paths_gegis(ssp_parentfolder, config):
...
-    load_paths,
-    regions,
-    admin_shapes,
-    countries,
-    scale,
-):
+def attach_load(n, demand_profiles):
+
     """
     Add load to the network and distributes them according GDP and population.
```
Comment (pz-max) [inline scripts/add_electricity.py:232]: Please check all the docstrings. This one needs to be adjusted.

File: scripts/add_electricity.py [hunk 23]
```diff
@@ -827,12 +762,10 @@ def add_nice_carrier_names(n, config):
     Nyears = n.snapshot_weightings.objective.sum() / 8760.0
 
     # Snakemake imports:
-    regions = snakemake.input.regions
-
-    load_paths = snakemake.input["load"]
     countries = snakemake.config["countries"]
     admin_shapes = snakemake.input.gadm_shapes
     scale = snakemake.config["load_options"]["scale"]
```
Comment (pz-max) [inline scripts/add_electricity.py]: Please remove unnecessary entries... I think `scale` is not required here?
Please check the others as well.

File: scripts/build_demand_profiles.py [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    substation_lv_i = n.buses.index[n.buses["substation_lv"]]
+    regions = gpd.read_file(regions).set_index("name").reindex(substation_lv_i)
+
+    load_paths = load_paths
+    # Merge load .nc files: https://stackoverflow.com/questions/47226429/join-merge-multiple-netcdf-files-using-xarray
+    gegis_load = xr.open_mfdataset(load_paths, combine="nested")
+    gegis_load = gegis_load.to_dataframe().reset_index().set_index("time")
+    # filter load for analysed countries
+    gegis_load = gegis_load.loc[gegis_load.region_code.isin(countries)]
+    logger.info(f"Load data scaled with scalling factor {scale}.")
```
Comment (pz-max) [inline scripts/build_demand_profiles.py]: scaling instead scalling

File: scripts/build_demand_profiles.py [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+    start_date,
+    end_date,
+):
+    """
+    Add load to the network and distributes them according GDP and population.
+
+    Parameters
+    ----------
+    n : pypsa network
+    regions : .geojson
```
Comment (pz-max) [inline scripts/build_demand_profiles.py:110]: change the order of the description to align with the read in above (line 94-101)

File: scripts/build_demand_profiles.py [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+    admin_shapes : .geojson
+        contains subregional gdp, population and shape data
+    countries : list
+        List of countries that is config input
+    scale : float
+        The scale factor is multiplied with the load (1.3 = 30% more load)
+
+    Returns
+    -------
+    n : pypsa network
```
Comment (pz-max) [inline scripts/build_demand_profiles.py]: No `n` is the output. Please review all the docstring for correctness and edit if necessary

File: scripts/build_demand_profiles.py [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+        Now attached with load time series
+    """
+    substation_lv_i = n.buses.index[n.buses["substation_lv"]]
+    regions = gpd.read_file(regions).set_index("name").reindex(substation_lv_i)
+
```
Comment (pz-max) [inline scripts/build_demand_profiles.py]: remove empty line

**Review by pz-max — CHANGES_REQUESTED** [review]
Just some little comments on docstring fixes etc. Otherwise, it looks good.
This PR is the foundation for future developments that allow passing new demand time-series :)

Agree with @davide-f  that checking the objective is the same would be great.

File: scripts/build_demand_profiles.py (hunk 24 above)
Comment (DeniseGiub) [reply scripts/build_demand_profiles.py]: Corrected

File: scripts/build_demand_profiles.py (hunk 27 above)
Comment (DeniseGiub) [reply scripts/build_demand_profiles.py]: done

File: scripts/build_demand_profiles.py (hunk 26 above)
Comment (DeniseGiub) [reply scripts/build_demand_profiles.py]: I had forgot to review docstrings for the functions. Now I corrected them according to the changes made

File: scripts/build_demand_profiles.py (hunk 25 above)
Comment (DeniseGiub) [reply scripts/build_demand_profiles.py:110]: done

File: scripts/add_electricity.py (hunk 23 above)
Comment (DeniseGiub) [reply scripts/add_electricity.py]: I removed them all except demand_profiles, since they were not needed

**Review by pz-max — APPROVED** [review]

**Comment by pz-max** [comment]
Congrats @DeniseGiub on the next big contribution. This will be merged after the checks :)
